### PR TITLE
Add execution details to config outputs

### LINKED ISSUE #384: Add user/environment metadata to output config files
Something along the lines of: 
```json
{
    "dataset": {
        "name": "doitt_buildingfootprints",
        "version": "20230402",
        "acl": "public-read",
        "source": { ... },
        "destination": {  ...  },
        "info": { ... } # some metadata related to source dataset
    },
    "execution_details": {
        "type": "manual"
        "username": fvankrieken
    }
}
```
or 
```json
"execution_details": {
    "type": "ci"
    "url": "https://github.com/NYCPlanning/db-developments/actions/runs/4653217871/jobs/8233898623"
}
```

### PR DESCRIPTION
**Pull request by fvankrieken**
Draft at the moment due to issue described below, but would like review/input

Closes #384 

Outputs look like this currently

[action](https://github.com/NYCPlanning/db-data-library/actions/runs/4702034994):
https://nyc3.digitaloceanspaces.com/edm-recipes/datasets/doitt_buildingfootprints/20230414/config.json
```json
"execution_details": {
    "type": "ci",
    "dispatch_event": "workflow_dispatch",
    "url": "https://github.com/NYCPlanning/db-data-library/actions/runs/4702034994/",
    "job": "dataloading",
    "timestamp": "2023-04-14 12:13:14"
}
```
job refers to specific job within the workflow run. Theoretically this has a url too, which I wanted to have originally rather than the overall run url (like `https://github.com/NYCPlanning/db-data-library/actions/runs/4702034994/jobs/12345`), but the `job_id` is not actually exposed as a variable, only the job name. So went with above.

manual
``` json
"execution_details": {
    "type": "manual",
    "user": "Finn van Krieken",
    "timestamp": "2023-04-14 12:13:14"
}
```

This "manual" one is grabbing the git username, since dev container requires that we add an ssh key to authenticate with git, I think this shouldn't fail. It also turned out to be fairly non-trivial to get some sort of username of the actual user in the dev container, this seemed like the simplest solution. But now that I write this, maybe I should add a try/catch, where on a failure to find any of this info we could maybe fail the job, or save the `execution_details` as `"could not resolve"`/`"unknown"` or something like that.

Currently have edit to single-runner.yml so that it uses the scripts in the repo to do archiving rather than using [action-library-archive](https://github.com/NYCPlanning/action-library-archive). Don't know if I should merge as is, but we certainly need to discuss. My gut is that this project should never use the published action, since current repo should be 1. most up to date on master and 2. only way to test development.

My vote would be actions in this repo never use the published action, but each commit/push to main should re-publish image to make sure they stay in sync

So to summarize, would love input on
1. format/naming of fields being added to config files
2. thoughts around the ci workflows and where they're pulling their logic from

### CODE REVIEW

**Comment by fvankrieken** [comment]
There's a bit of ugliness in the single-runner file as written as well - duplicated logic from both dockerfile and the entrypoint script in https://github.com/NYCPlanning/action-library-archive. Need to think about a slightly more elegant solution

**Comment by fvankrieken** [comment]
And before merging I'm gonna go ahead and fix these test failures, though I have not touched anything related to them

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -7,17 +7,19 @@
...
-	"settings": {
-		"terminal.integrated.shell.linux": null
+	"customizations": {
+		"vscode": {
+			"settings": {},
+			// Add the IDs of extensions you want installed when the container is created.
+			"extensions": [
+				"ms-python.python"
+			]
+		}
```
Comment (fvankrieken) [inline .devcontainer/devcontainer.json:17]: Just making VSCode happy with the formatting. Also it doesn't seem to be happy with the shell part, but it doesn't seem like removal has done anything negative

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -7,17 +7,19 @@
...
 	},
-	// Add the IDs of extensions you want installed when the container is created.
-	"extensions": [
-		"ms-python.python"
-	],
 	// Use 'forwardPorts' to make a list of ports inside the container available locally.
 	// "forwardPorts": [],
 	// Adding id_rsa so that we can push to github from the dev container
-	"initializeCommand": "ssh-add $HOME/.ssh/id_rsa",
+	"initializeCommand": "ssh-add",
```
Comment (fvankrieken) [inline .devcontainer/devcontainer.json:22]: I think DM has seen this, but @mbh329 if you haven't, this is a more generic ssh-add that will add whatever credentials you have in .ssh

File: library/utils.py [hunk 3]
```diff
@@ -52,3 +54,22 @@ def format_url(path: str, subpath: str) -> str:
         return url.replace("s3://", "/vsis3/")
 
     return url
+
+def get_execution_details():
+    today = date.today().strftime("%Y-%m-%d")
```
Comment (fvankrieken) [inline library/utils.py:59]: Went with date rather than timestamp just because the timestamp itself is a little arbitrary depending on when in the process it's actually taken. But if we think it'd be more useful to have slightly more specific info, would be happy to use timestamp instead

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (damonmcc) [reply .devcontainer/devcontainer.json:17]: yea I've never noticed a benefit from having the shell setting in other `devcontainer.json` files either, glad to remove it!

File: library/utils.py (hunk 3 above)
Comment (damonmcc) [reply library/utils.py:59]: agree it can be arbitrary, but seems worth the timestamp to avoid conflicts/confusion if we run something multiple times a day (like when debugging)

hour:minute would probably be precise enough and still readable

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (mbh329) [reply .devcontainer/devcontainer.json:17]: I think this is probably just a symptom of the old "copy and paste" and that probably persists in a lot of repos. Always great to simplify the  devcontainer.json file

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (mbh329) [reply .devcontainer/devcontainer.json:22]: I have not and thanks for flagging, should the comment above be updated as well?

File: .github/workflows/single-runner.yml [hunk 4]
```diff
@@ -8,31 +8,39 @@ on:
...
-      - uses: actions/checkout@v2.3.4
-      - uses: NYCPlanning/action-library-archive@v1.1
-        name: Archive ${{ github.event.inputs.dataset }}
-        with:
-          name: ${{ github.event.inputs.dataset }}
-          latest: ${{ github.event.inputs.latest }}
-          version: ${{ github.event.inputs.version }}
-          s3: true
-          compress: true
-          output_format: pgdump shapefile csv
```
Comment (mbh329) [inline .github/workflows/single-runner.yml:38]: What type of file formats does data library dump to Digital Ocean with the latest changes of removing the specification of the `output_format`?

File: .github/workflows/single-runner.yml [hunk 5]
```diff
@@ -8,31 +8,39 @@ on:
...
+      run: |
+        docker load --input /tmp/myimage.tar
+    - name: Archive ${{ github.event.inputs.dataset }}
+      env:
+        latest: ${{ github.event.inputs.latest == 'true' && '--latest' || ' ' }}
+        version: ${{ github.event.inputs.version || ' ' }}
+      run: |
+        echo $latest
+        echo $version
+        poetry run library archive --name ${{ github.event.inputs.dataset }} -o pgdump --s3 --compress $latest $version
```
Comment (mbh329) [inline .github/workflows/single-runner.yml:43]: I don't think we always want just the compressed files, what type of output files are pushed to DO (for a specific dataset) when the `single-runner` workflow is run?

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (damonmcc) [reply .devcontainer/devcontainer.json:22]: true. the one I wrote recently was 
```
# Add local SSH private keys in order to push to github from the dev container
```

File: .github/workflows/single-runner.yml (hunk 5 above)
Comment (mbh329) [reply .github/workflows/single-runner.yml:43]: ahh answer is in the write up

File: .github/workflows/single-runner.yml [hunk 6]
```diff
@@ -8,31 +8,39 @@ on:
...
-        with:
-          name: ${{ github.event.inputs.dataset }}
-          latest: ${{ github.event.inputs.latest }}
-          version: ${{ github.event.inputs.version }}
-          s3: true
-          compress: true
-          output_format: pgdump shapefile csv
\ No newline at end of file
+    - name: Setup Docker Buildx
+      uses: docker/setup-buildx-action@v1
```
Comment (mbh329) [inline .github/workflows/single-runner.yml]: What is this doing?

Comment (fvankrieken) [reply .github/workflows/single-runner.yml]: Removed, was part of docker testing

File: .github/workflows/single-runner.yml (hunk 5 above)
Comment (damonmcc) [reply .github/workflows/single-runner.yml]: @fvankrieken 

sorry I'm not seeing the answer in the write up. just flagging cuz it looks like this changes the `single-runner` output from the 3 formats `pgdump shapefile csv` to only the compressed `pgdump`

File: .github/workflows/single-runner.yml [hunk 7]
```diff
@@ -8,31 +8,39 @@ on:
...
       version:
         description: "The version of the dataset (i.e. 22v2, 21C) if needed (optional)"
         required: false
 
-
-
 jobs:
-  dataloading:
+  setup:
+    uses: ./.github/workflows/setup.yml
```
Comment (damonmcc) [inline .github/workflows/single-runner.yml]: all uses of a file from the repo in a workflow step _might_ first require `- uses: actions/checkout@v3`

but if it works without it then nvm!

File: library/utils.py [hunk 8]
```diff
@@ -52,3 +55,25 @@ def format_url(path: str, subpath: str) -> str:
         return url.replace("s3://", "/vsis3/")
 
     return url
+
+def get_execution_details():
+    def try_func(func):
+        try: return func()
+        except: return "could not parse"
```
Comment (damonmcc) [inline library/utils.py:62]: maybe this function + lambda isn't needed if all  uses of `os.environ["ENV_VAR"]` were replaced with something like `os.environ.get("ENV_VAR", "could not parse")` ?

that `dict.get("ENV_VAR")` approach let's you set a default value if the key isn't found, rather than the `KeyError` we get when using `dict["ENV_VAR"]`

File: .github/workflows/update_image.yml [hunk 9]
```diff
@@ -7,9 +7,12 @@ on:
         type: choice
         description: Name of release
         required: true
+        default: ubuntu-latest
```
Comment (mbh329) [inline .github/workflows/update_image.yml:10]: I am assuming this pulls in the 22.04 LTS version of Ubuntu, do we want to default to the latest version or pin the 20.02 (I might also be misunderstanding what this is doing)?

Comment (fvankrieken) [reply .github/workflows/update_image.yml:10]: So actually, this is how we store this docker image, the names are ubuntu-staging and ubuntu-latest

File: library/utils.py (hunk 8 above)
Comment (fvankrieken) [reply library/utils.py:62]: Yeah, the main reason for the alternative was the one row with 3 or 4 os.environ['thing'] and to just use the same wrapper for the function getting git username. Happy to tweak

File: .github/workflows/single-runner.yml (hunk 7 above)
Comment (fvankrieken) [reply .github/workflows/single-runner.yml]: This was reverted

File: .github/workflows/update_image.yml [hunk 10]
```diff
@@ -29,3 +32,59 @@ jobs:
...
       - name: Publish Tag
         run: docker push $image_tag
+
+  update_docs:
+    needs: update_image
+    runs-on: ubuntu-20.04
+    container: osgeo/gdal:ubuntu-small-3.6.1
+
+    steps:
+    - uses: actions/checkout@v2
```
Comment (mbh329) [inline .github/workflows/update_image.yml:42]: If we are going through the and doing maintenance, should we update `checkout@v2` to `v3`?

Comment (mbh329) [reply .github/workflows/update_image.yml:42]: https://github.com/actions/checkout

**Comment by fvankrieken** [comment]
> LGTM - one note/thought is whether it is worthwhile to update some of our regularly used actions?

Makes sense to me

File: library/utils.py (hunk 8 above)
Comment (fvankrieken) [reply library/utils.py:62]: Left for the two noted cases above, used .get() with default for the simpler cases

**Review by damonmcc — APPROVED** [review]
